pandas/core/dtypes/base.py

```diff
@@ -40,11 +43,19 @@ def __eq__(self, other):
         bool
         """
         if isinstance(other, compat.string_types):
-            return other == self.name
-        elif isinstance(other, type(self)):
-            return True
-        else:
-            return False
+            try:
+                other = self.construct_from_string(other)
+            except TypeError:
+                return False
+        if isinstance(other, type(self)):
+            return all(
+                getattr(self, attr) == getattr(other, attr)
+                for attr in self._metadata
+            )
```

Comment:
Corner case: other is an instance of a subclass of type(self) and self._metadata != other._metadata